nxt/sensor/hitechnic.py

```diff
@@ -459,9 +460,238 @@ def set_digital_modes(self, modes):
+        if len(boolean_list) != 8:
+            raise RuntimeError("List must be 8 booleans in length")
+        if reverse:
+            boolean_list.reverse()
+        output_byte = 0
+        for bit in range(0, 8):
+            output = (2 ** bit) * boolean_list[bit]
+            """
+            Debug prints
+            print("bit: {}".format(boolean_list[bit]))
+            print("value: {}".format(2 ** bit))
+            print("output: {}".format(output))
+            """
+            output_byte += output
+        return output_byte
+
+    def get_digital(self) -> dict[str, bool]:
+        """
+        Get digital input pins (D0-D7)
+        :return: Boolean list, LSB first.
+        """
+        digital_in = self.read_value("digital_in")[0]
+        boolean_list = self._byte_to_boolean_list(digital_in, reverse=True)
+        digital_in_map = {}
+        for x in range(0, 8):
+            digital_in_map["b{}".format(x)] = boolean_list[x]
+        return digital_in_map
+
+    def set_digital(self, pins: list[bool]):
+        """
+        Set digital output pins (D0-D7)
+        :param pins: boolean list (LSB first)
+        """
+        if len(pins) != 8:
+            raise RuntimeError("Need to specify list of 8 boolean values")
+        self.write_value("digital_out", [self._boolean_list_to_byte(pins)])
+
+    def set_digital_byte(self, integer: int, lsb=True):
+        """
+        Set digital output pins (D0-D7) from byte
+        :param integer: Byte (0-255 inclusive)
+        :param lsb: Whether to output in LSB order (0x01 = pin B0)
+        """
+        if not (0 <= integer <= 255):
+            raise RuntimeError("Integer must be in range of 0 to 255 inclusive")
+        self.set_digital(self._byte_to_boolean_list(integer, reverse=lsb))
+
+    def set_digital_modes(self, modes: list[bool]):
+        """
+        Set digital pin mode (D0-D7)
+        :param modes: boolean list (LSB first, True = Output, False = Input)
+        """
+        if len(modes) != 8:
+            raise RuntimeError("Need to specify list of 8 boolean values")
+        self.write_value("digital_dir", [self._boolean_list_to_byte(modes)])
+
+    def set_digital_modes_byte(self, mode_int: int, lsb=True):
+        """
+        Set digital output pins mode (D0-D7) from byte
+        :param mode mode_int: Byte (0-255 inclusive)
+        :param lsb: Whether to output in LSB order (0x01 = pin B0)
+        """
+        if not (0 <= mode_int <= 255):
+            raise RuntimeError("Integer must be in range of 0 to 255 inclusive")
+        self.set_digital_modes(self._byte_to_boolean_list(mode_int, reverse=lsb))
+
+    def set_strobe_output(self, mode_int: int):
+        """
+        Strobe output ??
+        It's unclear what the Strobe output does. It appears to just output like the output pins do.
+        S0 = 1, S1 = 2 S2 = 4 S3 = 8 RD = 16 WR = 32 (64 and 128 have seemingly no meaning??)
```

Comment:
Is it something like having RD or WR output a pulse? To drive a 4 bit LCD maybe?

Reply:
Yes, some kind of 4 bit bus…

Reply:
I'm going to have to test that functionality again; I don't recall seeing any pulses on the scope.. they acted like any other logical output and didn't change state rapidly/without my doing so. The 8 pin digital inputs/ouputs behave very similarly to this, except this just doesn't have the mode or input byte.

Reply:
OK. Did some more testing.

The strobe output behavior is a bit odd.

The strobe does not appear to strobe any particular data, but if the strobe bits are on, and I send a digital read or write to the B0-B7 pins, it will send a 1us pulse on the corresponding strobe lines. Setting the strobe bits doesn't appear to affect what the strobe strobes, it's always a 500ns pulse.

I'm going to guess that this could be useful for sending a strobe pulse for some use cases once data is in the B0-B7 and perhaps S0-S3 lines for like a shift register, but would like to mention that at least on my Linux testing box, if I use the following code:

```python
# setup
pro.set_digital_modes_byte(0xFF)
pro.set_digital_byte(0b01010101)
# actual test
pro.set_strobe_output(0b00111010)
pro.get_digital()
pro.set_digital_byte(0b01010101)
pro.set_strobe_output(0x00)
```
The part where the strobe output is set raises the RD and WR lines from 0V to ~5V, there's an almost 10ms delay until the strobe of the read line, another almost exactly 10ms delay until the strobe of the write line, and then 10ms more until the output goes low.

Reply:
(forgot to mention - the strobe is a logical low pulse when the line is drawn high by setting the strobe bits; will test if the strobe is inverted if the bit isn't set)

Reply:
Yep. Results in line low but being pulled high if strobe bits 32 and 64 are low.